[doc] Improve developer installation documentation (#1343)
taichi-dev/taichi: docs/dev_install.rst

  .. code-block:: bash

>     python3 -m pip install --user setuptools wheel astor pybind11 Pillow

Reviewer: We no longer need `dill` here?

Author: ```suggestion
    python3 -m pip install --user setuptools wheel astor pybind11 Pillow dill
```

    python3 -m pip install --user pytest pytest-rerunfailures pytest-xdist yapf
    python3 -m pip install --user numpy GitPython coverage colorama autograd

- Make sure you have ``clang`` with version >= 7

  * On Windows: Download ``clang-8`` via `this link <https://releases.llvm.org/8.0.0/LLVM-8.0.0-win64.exe>`_.
    Make sure you add the ``bin`` folder containing ``clang.exe`` to the ``PATH`` environment variable.

  * On OS X: you don't need to do anything.

  * On Ubuntu, execute ``sudo apt install clang-8``.
  

>   * On Arch Linux, execute ``sudo pacman -S clang``. (this will install ``clang-10``, which should work as well).

Reviewer: ```suggestion
  * On Arch Linux, execute ``sudo pacman -S clang``. (This will install ``clang-10``, which should work too).
```

- On **Arch Linux**, you can easily install CUDA via ``pacman -S cuda`` without downloading the installer manually.

.. _dev_env_settings:

Setting up Taichi for development
---------------------------------

> - Set up environment variables for Taichi:

Reviewer 2: To set up environment variables for Taichi:

  * On Linux / OS X, please add the following script to your ``~/.bashrc``:

    .. code-block:: bash

      export TAICHI_REPO_DIR=/home/XXX/taichi  # Path to your taichi repository
      export PYTHONPATH=$TAICHI_REPO_DIR/python:$PYTHONPATH
      export PATH=$TAICHI_REPO_DIR/bin:$PATH
      # export PATH=/opt/llvm/bin:$PATH  # Uncomment if your llvm-8 or clang-8 is in /opt

    Then execute ``source ~/.bashrc`` to reload shell config.

  * On Windows, please add these variables by accessing your system settings:

>     1. Add ``TAICHI_REPO_DIR`` whose value is path to your taichi repository so that Taichi knows you're a developer.

Reviewer 2: ... whose value is the path to ...

Setting up Taichi for development
---------------------------------

- Add the following script to your ``~/.bashrc``:

  .. code-block:: bash

    export TAICHI_REPO_DIR=/home/XXX/taichi  # Path to your taichi repository
    export PYTHONPATH=$TAICHI_REPO_DIR/python:$PYTHONPATH
    export PATH=$TAICHI_REPO_DIR/bin:$PATH
    # export PATH=/opt/llvm/bin:$PATH # Uncomment if your llvm-8 or clang-8 is in /opt

- Execute ``source ~/.bashrc`` to reload shell config.

> 

Reviewer: Could this be confusing given `/home/XXX/taichi` doesn't exist yet?

Reviewer: I realized we didn't really mention setting these variables on Windows. Maybe we should mention it here?

    mkdir build
    cd build
    cmake ..
    # On Linux / OS X, if you do not set clang as the default compiler
    # use the line below:
    #   cmake .. -DCMAKE_CXX_COMPILER=clang-8
    #
    # Alternatively, if you would like to set clang as the default compiler
    # On Unix CMake honors environment variables $CC and $CXX upon deciding which C and C++ compilers to use
    make -j 8

> - Check out ``examples`` for runnable examples. Run them with command like ``python3 examples/mpm128.py``.

Reviewer 2: ... with commands like ...

- Execute ``python3 -m taichi test`` to run all the tests. It may take up to 5 minutes to run all tests.

Troubleshooting Developer Installation
--------------------------------------

- If ``make`` fails to compile and reports ``fatal error: 'spdlog/XXX.h' file not found``,
  please try run ``git submodule update --init --recursive --depth=1``.

> - If importing Taichi cause

Reviewer 2: ... causes

  .. code-block:: none

      FileNotFoundError: [Errno 2] No such file or directory: '/root/taichi/python/taichi/core/../lib/taichi_core.so' -> '/root/taichi/python/taichi/core/../lib/libtaichi_core.so'``

  Please try adding ``TAICHI_REPO_DIR`` to environment variables, see :ref:`dev_env_settings`.

> - If the build succeed, but running any Taichi code will result in errors like ``Bitcode file (/tmp/taichi-tero94pl/runtime//runtime_x64.bc) not found``,

Reviewer 2: If the build succeeded but running any Taichi code results in errors like ...

    #   cmake .. -DCUDA_VERSION=10.0 -DTI_WITH_CUDA:BOOL=True
    make -j 8

- Execute ``python3 -m taichi test`` to run all the tests. It may take up to 5 minutes to run all tests.
- Check out ``examples`` for runnable examples. Run them with ``python3``.

> Developer's Troubleshooting
> ---------------------------

Reviewer: ```suggestion
Troubleshooting Developer Installation
--------------------------------------
```

> - If ``make`` failed to compile and reporting ``fatal error: 'spdlog/XXX.h' file not found``,

Reviewer: ```suggestion
- If ``make`` fails to compile and reports ``fatal error: 'spdlog/XXX.h' file not found``,
```

  please try run ``git submodule init --update --recursive --depth=1``.

> - If ``cmake`` failed and reporting ``clang and MSVC are the only supported compilers for Taichi compiler development``,
>   please make sure you have installed clang with version >= 7, and build with ``cmake -DCMAKE_CXX_COMPILER=clang``.

Reviewer: ```suggestion
```

This seems to duplicate the CMakeList.txt message?

- If importing Taichi cause ``FileNotFoundError: [Errno 2] No such file or directory: '/root/taichi/python/taichi/core/../lib/taichi_core.so' -> '/root/taichi/python/taichi/core/../lib/libtaichi_core.so'``, please:

  * On Windows, add a environment variable called ``TAICHI_REPO_DIR`` with value ``C:\\path\\to\\taichi`` (your Taichi repo path), this may be either done in your IDE configuation panel or system settings.

>   * On Linux / OS X, add ``export TAICHI_REPO_DIR=/home/XXX/taichi`` to your ``~/.bashrc`` and restart ``bash``.

Reviewer: Maybe we should just refer the readers to the environment variable settings instead of repeating ourselves?

> - If the build succeed, but running any Taichi code will result in an error like ``Bitcode file (/tmp/taichi-tero94pl/runtime//runtime_x64.bc) not found``,

Reviewer: ```suggestion
- If the build succeed, but running any Taichi code will result in errors like ``Bitcode file (/tmp/taichi-tero94pl/runtime//runtime_x64.bc) not found``,
```

  please make sure ``clang`` is in your ``PATH``:

  .. code-block:: bash
  
      clang --version
      # version should be >= 7
      
  and our **Taichi configured** ``llvm-as``:

  .. code-block:: bash
  
      llvm-as --version

>       # version should be 8.0.1 or 10.0.0

Reviewer: ```suggestion
      # version should be >= 8
```

      which llvm-as
      # should be /usr/local/bin/llvm-as or /opt/XXX/bin/llvm-as, which is our configured installation
      
  If not, please install ``clang`` and **build LLVM from source** with instructions above in `:ref:dev_install`, then add their path to environment variable ``PATH``.

> - If you encountered other issues, feel free to report it by `opening an issue on GitHub <https://github.com/taichi-dev/taichi/issues/new?labels=potential+bug&template=bug_report.md>`_, we are willing to help!

Reviewer: ```suggestion
- If you encountered other issues, feel free to report by `opening an issue on GitHub <https://github.com/taichi-dev/taichi/issues/new?labels=potential+bug&template=bug_report.md>`_. We are willing to help!
```

- Execute ``python3 -m taichi test`` to run all the tests. It may take up to 5 minutes to run all tests.
- Check out ``examples`` for runnable examples. Run them with ``python3``.

Developer's Troubleshooting
---------------------------

- If ``make`` failed to compile and reporting ``fatal error: 'spdlog/XXX.h' file not found``,
  please try run ``git submodule init --update --recursive --depth=1``.

- If ``cmake`` failed to generate and reporting ``clang and MSVC are the only supported compilers for Taichi compiler development``,
  please make sure you have installed clang with version >= 7, and add ``export CMAKE_CXX_COMPILER=/path/to/clang`` to ``.bashrc``.

- If importing Taichi cause ``FileNotFoundError: [Errno 2] No such file or directory: '/root/taichi/python/taichi/core/../lib/taichi_core.so' -> '/root/taichi/python/taichi/core/../lib/libtaichi_core.so'``, please:

  * On Windows, add a environment variable called ``TAICHI_REPO_DIR`` with value ``/home/XXX/taichi`` (Taichi repo path), this may be either done in your IDE configuation panel or system settings.
  * On Linux / OS X, add ``export TAICHI_REPO_DIR=/home/XXX/taichi`` to your ``~/.bashrc`` and restart ``bash``.

> - If the build succeed, but running a Taichi example result in an error, please make sure ``clang`` is in your ``PATH``:

Author: ```suggestion
- If the build succeed, but running a Taichi example result in an error like ``Bitcode file (/tmp/taichi-tero94pl/runtime//runtime_x64.bc) not found``,
  please make sure ``clang`` is in your ``PATH``:
```

>     #   cmake .. -DCUDA_VERSION=10.0 -DTI_WITH_CUDA:BOOL=True

Author: Btw, I thought this is no longer needed thanks to `cuda_driver.h`?

Reviewer: You are right!
